### ISSUE TITLE
Get-Content leaks memory, even when discarding output, causing visible slowdown

### ISSUE THREAD

**Issue by dkaszews** [PowerShell/PowerShell#20892]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

Processing text files, I have noticed that PowerShell is slowing down, even when discarding the output (only checking for errors). Looking at Task Manager, the memory usage keeps going up and is never released. It may go down couple MBs as other values are garbage collected, but does not go to original levels.

### Expected behavior

```console
> function Measure-Memory { (Get-Process -Id $PID).PagedMemorySize / 1MB }  # Seems to match Task Manager the closest
> function Test($File) { (Measure-Command { Get-Content $File | Out-Null }).TotalSeconds; Measure-Memory }
> 1..3 | %{ Test 100mb.txt }
# Dependent on environment but stable
1.0
100
1.0
100
1.0
100
```

### Actual behavior

```console
# Runtime goes up by 0.5s, memory usage goes up by 25MB
1.0
100
1.5
125
2.0
150
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.0
PSEdition                      Core
GitCommitId                    7.4.0
OS                             Microsoft Windows 10.0.19044
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by 237dmitry** [comment]
I could not reproduce on Linux:

```powershell
dd if=/dev/zero of=./file.txt bs=100M count=1

function Measure-Memory { sls 'Rss' /proc/$pid/status -raw }
function Test($File) { (Measure-Command { Get-Content $File | Out-Null }).TotalSeconds; Measure-Memory;'' }
1..3 | % { Test file.txt }
```
output:

```
0,1113467
VmRSS:	 2289764 kB
RssAnon:	 2181496 kB
RssFile:	   76524 kB
RssShmem:	   31744 kB

0,0759255
VmRSS:	 2289764 kB
RssAnon:	 2181496 kB
RssFile:	   76524 kB
RssShmem:	   31744 kB

0,0762557
VmRSS:	 2289764 kB
RssAnon:	 2181496 kB
RssFile:	   76524 kB
RssShmem:	   31744 kB
```
`PagedMemorySize` or `PM` (PagedMemorySize64) returned `0`

```powershell
$ (Get-Process -Id $PID).PagedMemorySize
0
```

Perhaps I looked at the wrong parameters. Entire output of `/proc/$pid/status`
```powershell
 $ gc /proc/$pid/status
Name:	pwsh
Umask:	0022
State:	S (sleeping)
Tgid:	7602
Ngid:	0
Pid:	7602
PPid:	7592
TracerPid:	0
Uid:	1000	1000	1000	1000
Gid:	1000	1000	1000	1000
FDSize:	256
Groups:	3 90 96 98 983 985 986 988 990 991 995 998 1000 
NStgid:	7602
NSpid:	7602
NSpgid:	7602
NSsid:	7602
Kthread:	0
VmPeak:	274629696 kB
VmSize:	274588656 kB
VmLck:	       0 kB
VmPin:	       0 kB
VmHWM:	 2289764 kB
VmRSS:	 1679228 kB
RssAnon:	 1570960 kB
RssFile:	   76524 kB
RssShmem:	   31744 kB
VmData:	 1770356 kB
VmStk:	     132 kB
VmExe:	      44 kB
VmLib:	   76148 kB
VmPTE:	    3788 kB
VmSwap:	       0 kB
HugetlbPages:	       0 kB
CoreDumping:	0
THP_enabled:	1
untag_mask:	0xffffffffffffffff
Threads:	14
SigQ:	0/127243
SigPnd:	0000000000000000
ShdPnd:	0000000000000000
SigBlk:	0000000000000000
SigIgn:	0000000000001000
SigCgt:	00000003080344fe
CapInh:	0000000800000000
CapPrm:	0000000000002000
CapEff:	0000000000002000
CapBnd:	000001ffffffffff
CapAmb:	0000000000000000
NoNewPrivs:	0
Seccomp:	0
Seccomp_filters:	0
Speculation_Store_Bypass:	thread vulnerable
SpeculationIndirectBranch:	conditional enabled
Cpus_allowed:	ff
Cpus_allowed_list:	0-7
Mems_allowed:	00000001
Mems_allowed_list:	0
voluntary_ctxt_switches:	488
nonvoluntary_ctxt_switches:	14
x86_Thread_features:	
x86_Thread_features_locked:	
```

**Comment by SteveL-MSFT** [comment]
If this repros for you, can you try: `[system.gc]::Collect()`?  Discarded output wouldn't be freed immediately as it's left to the garbage collector.

**Comment by dkaszews** [comment]
Tried that, no effect. I know it wouldn't be freed immediately, but there should be enough pressure on GC that it gets freed in reasonable time, and I saw it hit couple GBs over the course of an hour of normal use (stuff like `Get-Content bigfile.txt | MyScript.ps1 -ErrorAction Stop | Out-Null`.

I can reproduce it on both my corporate and private machine.

**Comment by dkaszews** [comment]
For testing, you should use a file with many lines, which `dd /dev/zero` does not produces. I attached my test payload which is just a single GUID duplicated thousands of times

[file.zip](https://github.com/PowerShell/PowerShell/files/13639639/file.zip)

**Comment by StevenBucher98** [comment]
Looks like I am able to reproduce on Windows: 
![image](https://github.com/PowerShell/PowerShell/assets/20605456/f0520ff8-1c5a-4664-8409-a944df6b9f81)

**Comment by iSazonov** [comment]
It's all more like a measurement fallacy. I ran this test and I can see that memory is being eaten. But after 5-15 minutes of waiting (no action in pwsh) Task Manager shows reduced size for the pwsh process.
It is also unclear why you think that PagedMemorySize is an indicator of memory leaks. This parameter shows only the size of virtual memory of the process. A process may declare that it needs 1 TB of virtual memory but actually use 1 MB of physical memory.

**Comment by dkaszews** [comment]
@iSazonov Don't really care about the virtual memory being shown, just the slowdown which is very noticeable and makes it impossible to work on larger files

**Comment by dkaszews** [comment]
But you are right that allocating even couple GBs should not be an issue as long as you don't start hitting the pagefile. Rather this shows that PowerShell itself is slowing down the more objects it creates, as if it was constantly traversing some structure as it gets bigger.

Will check older versions, as I don't remember this being an issue around 2 years back which was last time I worked on files this large.

**Comment by dkaszews** [comment]
New findings:
* Does reproduce with `-NoProfile`
* Does reproduce on the same machine under WSL
* Does reproduce on all of 7.0.0, 7.1.0, 7.2.0, 7.3.0, 7.4.0 with roughly the same values (in fact newer versions are ever so slightly faster)

Repeating experiment after ~30 minutes shows same results as on fresh PowerShell - the times and memory still go up, but their values restart. This suggests not a leak per se, but a lazy GC that somehow significantly impacts runtime.

For anyone trying to repro, please use the payload I linked above, as the repro seems more linked to number of string objects piped and hence number of lines in the file, not its raw size.

The weirdest thing is that if first run takes `N` seconds, then second one takes `2N`, third one `3N` etc - it's roughly linear on number of runs. I verified it with external stopwatch that it's not some odd measurement error. The memory usage is not linear even accounting its initial offset, the delta decreases between runs.

**Comment by iSazonov** [comment]
After playing around with it and taking some measurements I don't see any memory leaks or handles.
There are two things worth noting (although they have been mentioned many times in the past)
1. The memory manager in .Net works differently than in the .Net Framework. And if Windows PowerShell has been extremely optimized for the .Net Framework, this code may not work optimally for .Net and it is worth making optimizations for .Net.
2. A test from this issue with different parameters shows that objects move quickly from Gen 0 to Gen 1 and then to Gen 2, and sometimes even continue from Gen 2 to LOH. While Gen 0 and Gen 1 are cleared very quickly, the accumulation of objects occurs in Gen 2 (and/or LOH) which takes several minutes rather than fractions of a second to clear. As a result, we are seeing unnecessary bloat in the process. If we increase the number of cycles from 3 to 300 (so that the test script is executed for 5-10 minutes), we can see that Gen 2/LOH is cleaned every few minutes and the process gets smaller (which is also visible in Task Manager).
It should be noted that Windows PowerShell has the same problem with bloat. This suggests that it is worth considering how to make memory handling better in PowerShell. Perhaps you can somehow tweak GC or insert hints for it into the pwsh code.

**Comment by dkaszews** [comment]
@iSazonov Thank you for detailed analysis. Any clue why this only reproduces on Windows and not Linux? Different .NET versions?

To reiterate, leaking couple hundred MBs on a 16GB machine should not affect performance so much. Maybe it's not an issue in memory itself, but some action that is only done in destructor that could be done in immediately in `Dispose`?

I'm more of a C++ guy, but I believe this kind of behavior could be modeled by the following:

```csharp
class Foo {
    private static int counter = 0;
    public Foo() { Thread.Sleep(counter++); }
    public ~Foo() { counter--; }
    public static void Test() {
        foreach (var i in Enumerable.Range(0, 100)) {
            var foo = new Foo();
        }
    }
}

Foo.Test();  // ~5s
Foo.Test();  // ~15s
GC.Collect();  // Force trigger all dtors to reset counter
Foo.Test();  // ~5s again
```

**Comment by iSazonov** [comment]
> Any clue why this only reproduces on Windows and not Linux? Different .NET versions?

Obviously, the memory manager/GC should depend on how the OS manages memory to be as fast as possible.
---

There is no point in reasoning in general, much less thinking in terms of C++, since the performance problem is always concrete. In this case it is how the .Net memory manager/GC relates to the peculiarities of a particular application (pwsh).

**Comment by iSazonov** [comment]
```
 [GC]::GetGCMemoryInfo()

HighMemoryLoadThresholdBytes : 3775872614
MemoryLoadBytes              : 3859780894
TotalAvailableMemoryBytes    : 4195414016
HeapSizeBytes                : 439345336
FragmentedBytes              : 33798736
Index                        : 437
Generation                   : 1
Compacted                    : True
Concurrent                   : False
TotalCommittedBytes          : 446423040
PromotedBytes                : 3892024
PinnedObjectsCount           : 0
FinalizationPendingCount     : 44
PauseDurations               :
PauseTimePercentage          : 0
GenerationInfo               :
```

After waiting some minutes I still see `FinalizationPendingCount     : 44` - I guess it is a root of the issue.

**Comment by stevenebutler** [comment]
On Windows, Get-Content appears to be quite slow compared to `[System.IO.File]::GetAllLines()` if you are just reading lines from a file.

First, every line output by Get-Content is added to a list, then returned as an array by the file system provider. Given the return type is IList it probably doesn't need to convert the list to an array, but this doesn't add any appreciable overhead to overall time.

The single item is then taken out of the array and wrapped in a PSObject, and finally the following note properties are added to
the PSObject to show where it came from. All lines have the same cached properties added (except ReadCount, which seems to hold the line number for each line read) but it will still take some time to add all the note properties to all of the PsObjects.

```
PSChildName
PSDrive
PSParentPath
PSPath
PSProvider
ReadCount
```
Since this happens to every single line in the file when using Get-Content (without -raw) this will be expensive for a large number of lines.

I noticed after running get-content on the provided file a few times that even `[system.io.File]::ReadAllLines()` slows down intermittently between 8s to 16s but in a new pwsh without having used Get-Content, it is consistently taking only around 1s to  read the file on my machine so .

When I unset variables assigned from the Get-Content then ran [GC]::Collect(), the time to run `ReadAllLines` went back to about 1 to 1.5s.

I ran Get-Content on your test file on Windows PowerShell 5.1 for comparison and it didn't seem any faster than PowerShell 7.4.0.

On 7.4.0 Running measure-command { $lines = get-content "file.txt" } is taking a long time for me:
Run 1: 55650 ms
Run 2: 180765  ms
(it seems super slow now after running some tests in other shells)

On latest master, I edited `ContentCommandBase.cs` and removed all the logic that wraps in a psobject and adds the property and the same file now takes:
Run 1: 5652 ms
Run 2: 8894 ms
Run 3: 6031 ms

After closing the 7.4.0 shell, which still had all the memory allocated from the previous tests the time in my patched version dropped further:
Run 4: 2526 ms
Run 5: 3414 ms

I will note that with the patched pwsh, I see it is using over 2GB to store the result (active private in task manager).
On Pwsh 7.4.0, it is using 4.8GB to store the result. Running [gc]::collect() only freed up 0.1 GB of this (the variable is still assigned though).

In both cases, this seems like a lot of memory to process a 100 MB file.

**Comment by iSazonov** [comment]
It's not surprising that it consumes a lot of memory if you store the result in a variable. That's the price we pay for the magical things PowerShell does. There is no way to remove these enhanced properties completely. (Although perhaps some optimizations can be made there (if not already done)).
I am more concerned about the scenario without saving the read data when it is just processed in pipeline and not used anymore. In this case PowerShell should free memory very fast and not consume more memory than needed to process a single line.

**Comment by dkaszews** [comment]
@daxian-dbw How is this answered? We have barely inched towards the root cause and are nowhere near a fix.

**Comment by dkaszews** [comment]
Please reopen

**Comment by dkaszews** [comment]
Seems to have been resolved in 7.4.3
